### PR TITLE
feat: Added APDS9960 Gesture, Color, Proximity and Ambient Light Sensor

### PR DESCRIPTION
**Pull request by AsCress**
Adds support for the **APDS9960 Gesture, Color, Proximity and Ambient Light Sensor**.

The **APDS9960** is a versatile digital, RGB, ambient light, proximity, and gesture sensor.
Some of its features: -
**1. RGB Color Sensing:** It can accurately measure red, green, and blue (RGB) light levels, allowing it to detect and differentiate between different colors in the environment.
**2. Ambient Light Sensing:** It includes an ambient light sensor capable of measuring the intensity of ambient light.
**3. Proximity Sensing:** It can detect the presence of objects or obstacles in close proximity to the sensor. It emits infrared (IR) light and measures the reflection to determine the distance between the sensor and nearby objects.
**4. Gesture Detection:** One of its standout features is its ability to recognize a variety of hand gestures, such as swipe, flick, and circle motions.

## Changes 
- The UI for the sensor (_sensor_apds9960.xml_) is developed akin to the UI for the existing sensors, providing raw data and graphical plots in a clear and user-friendly manner.
- The sensor activity (_SensorAPDS9960.java_) follows on a similar pattern as the already present sensors.
- A dedicated communication library for the sensor (_APDS9960.java_) is developed by porting the [adafruit](https://github.com/adafruit/Adafruit_CircuitPython_APDS9960/tree/main) library for the sensor to Java.

## Screenshots / Recordings  
Here is a screen recording demonstrating the functioning: -

https://github.com/user-attachments/assets/0f172926-7133-4bd5-8090-94dcf85f9014

**Checklist**: 
- [x] **No hard coding**: I have used resources from `strings.xml`, `dimens.xml` and `colors.xml` without hard coding any value.
- [x] **No end of file edits**: No modifications done at end of resource files `strings.xml`, `dimens.xml` or `colors.xml`.
- [x] **Code reformatting**: I have reformatted code and fixed indentation in every file included in this pull request.
- [x] **No extra space**: My code does not contain any extra lines or extra spaces than the ones that are necessary.

**Note:-**
This has been tested with the sensor.
@marcnause Could you please also test this and confirm if this works fine ?

## Summary by Sourcery

Introduce support for the APDS9960 sensor, including gesture, color, proximity, and ambient light sensing. Develop a new communication library and user interface for the sensor, enhancing the application's sensor capabilities.

New Features:
- Add support for the APDS9960 sensor, enabling gesture, color, proximity, and ambient light sensing capabilities.

Enhancements:
- Develop a dedicated communication library for the APDS9960 sensor by porting the Adafruit library to Java.
- Implement a user interface for the APDS9960 sensor, providing raw data and graphical plots in a user-friendly manner.

### CODE REVIEW

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+        setSensorDock();
+        sensorDock.setVisibility(View.VISIBLE);
+
+
+        scienceLab = ScienceLabCommon.scienceLab;
+        I2C i2c = scienceLab.i2c;
+        try {
+            sensorAPDS9960 = new APDS9960(i2c, scienceLab);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (marcnause) [inline app/src/main/java/io/pslab/sensors/SensorAPDS9960.java]: I think writing to Log would be better.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+                        if (flag == 0) {
+                            startTime = System.currentTimeMillis();
+                            flag = 1;
+                        }
+
+                        synchronized (lock) {
+                            try {
+                                lock.wait();
+                            } catch (InterruptedException e) {
+                                e.printStackTrace();
```
Comment (marcnause) [inline app/src/main/java/io/pslab/sensors/SensorAPDS9960.java]: Same as above.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+                            }
+                        }
+                        try {
+                            Thread.sleep(timeGap);
+                        } catch (InterruptedException e) {
+                            e.printStackTrace();
```
Comment (marcnause) [inline app/src/main/java/io/pslab/sensors/SensorAPDS9960.java]: Same as above.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+                if (sensorAPDS9960 != null) {
+                    sensorAPDS9960.enableGesture(false);
+                    sensorAPDS9960.enableColor(true);
+                    sensorAPDS9960.enableProximity(true);
+                    dataAPDS9960Color = sensorAPDS9960.getColorData();
+                    dataAPDS9960Lux = (-0.32466 * dataAPDS9960Color[0]) + (1.57837 * dataAPDS9960Color[1]) + (-0.73191 * dataAPDS9960Color[2]);
+                    dataAPDS9960Proximity = sensorAPDS9960.getProximity();
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (marcnause) [inline app/src/main/java/io/pslab/sensors/SensorAPDS9960.java:320]: Same as above.

**Comment by marcnause** [comment]
I successfully tested the sensor (proximity, color and light sensing) with the v5 board and the v3.0.2 firmware. We agreed that gestures will be added in a future PR.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java (hunk 1 above)
Comment (AsCress) [reply app/src/main/java/io/pslab/sensors/SensorAPDS9960.java]: Replaced with logging.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java (hunk 2 above)
Comment (AsCress) [reply app/src/main/java/io/pslab/sensors/SensorAPDS9960.java]: Replaced with logging.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java (hunk 3 above)
Comment (AsCress) [reply app/src/main/java/io/pslab/sensors/SensorAPDS9960.java]: Replaced with logging.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java (hunk 4 above)
Comment (AsCress) [reply app/src/main/java/io/pslab/sensors/SensorAPDS9960.java:320]: Replaced with logging.

**Comment by AsCress** [comment]
@CloudyPadmal @marcnause @bessman 
This PR now adds support for gestures too using the APDS9960 !! 🥳🥳
Support for four basic (`LEFT`, `RIGHT`, `UP`, `DOWN`) gestures has been implemented as done in the [adafruit](https://github.com/adafruit/Adafruit_CircuitPython_APDS9960) driver for the sensor.
Since only one of the engines of the sensor, either the _gesture_ engine or the _proximity_ engine can work at a time, the features have been separated in the app.
Here is a screen recording demonstrating the complete functionality implemented for the sensor in the app: -

https://github.com/user-attachments/assets/e760f9f7-bf1e-40de-b0a5-ac82677694f1

@marcnause Could you also please test these features with the sensor and confirm that they work fine? 

P.S.- Please ignore the issues found out by static code analysis as they are complexity related and we can't do much to reduce the complexity when the driver functionality requires that (ported from the adafruit driver).

**Comment by marcnause** [comment]
Yay! 🥳 I tested the sensor again and I can confirm that the gesture detection is working fine with my sensor and is fun to use. 😉

I think that `getGesture` could be refactored to reduce the complexity (e.g. move block in the `while`-loop to an extra method), but I am nit sure that it would make the code more readable.

Would it be possible to increase the width of the `Spinner` for he mode selection? Also I think it would be easier for the user to understand the screen if the mode selection would be the first element of the screen. What do you think?

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+
+import io.pslab.DataFormatter;
+import io.pslab.R;
+import io.pslab.communication.ScienceLab;
+import io.pslab.communication.peripherals.I2C;
+import io.pslab.communication.sensors.APDS9960;
+import io.pslab.others.ScienceLabCommon;
+
+public class SensorAPDS9960 extends AppCompatActivity {
+    private final String TAG = this.getClass().getSimpleName();
```
Comment (marcnause) [inline app/src/main/java/io/pslab/sensors/SensorAPDS9960.java]:
```suggestion
    private static final String TAG = this.getClass().getSimpleName();
```
I know, that's nitpicking... 😉

**Comment by AsCress** [comment]
@marcnause Thanks for testing! Your suggestions regarding the layout are nice ones! I've updated the layout accordingly. Even if we refactor the `getGesture` method, we won't be able to reduce the N-Path complexity (I've already done quite a few refactoring by moving the code into other methods).

**Comment by marcnause** [comment]
@AsCress Thumbs up for the changed layout! I think it looks really good.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+
+import io.pslab.DataFormatter;
+import io.pslab.R;
+import io.pslab.communication.ScienceLab;
+import io.pslab.communication.peripherals.I2C;
+import io.pslab.communication.sensors.APDS9960;
+import io.pslab.others.ScienceLabCommon;
+
+public class SensorAPDS9960 extends AppCompatActivity {
+    private static final String TAG = "APDS9960";
```
Comment (marcnause) [inline app/src/main/java/io/pslab/sensors/SensorAPDS9960.java]: Sorry, I was not paying attention when I made the previous suggestion. I usually create TAGs like this:

```suggestion
    private static final String TAG = SensorAPDS9960.class.getSimpleName();
```

The nice thing about it is that if the class name us changed later, the TAG will be updated automatically.

File: app/src/main/java/io/pslab/sensors/SensorAPDS9960.java (hunk 4 above)
Comment (marcnause) [reply app/src/main/java/io/pslab/sensors/SensorAPDS9960.java:320]: Ooops! That change did not make it to your git commit. Could you please have a look at it again?

**Comment by AsCress** [comment]
@marcnause Is there something that you would like to discuss regarding this PR or maybe give it a final review ? This is ready to be merged from my side.

**Review by marcnause — CHANGES_REQUESTED** [review]

**Review by marcnause — APPROVED** [review]

**Pull request merged**